APM Synthtrace: Introduce new DSL using generators

Mpdreamz opened the pull request:
## Summary

This updates the [`@elastic/apm-synthtrace`](https://github.com/elastic/kibana/tree/main/packages/elastic-apm-synthtrace) in various ways.


* Updates the DSL to setup an async generator of events instead of a fully materialized array of the events.
* This allows the DSL to express a long range.
* Moves away from custom workers and custom partitioning to generate data for a long range.
* Feeds the async generator directly into [`client.helpers.bulk`](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/master/client-helpers.html#bulk-helper)

This PR also models stream processing explicitly through `StreamProcessor` which will emit the source event + additional derivative events 

```typescript
const javaService = apm.service('opbeans-java', 'production', 'java');
const javaInstance = javaService.instance('instance-1');
const eventStream = range
	  .interval('1m')
	  .rate(25)
	  .spans((timestamp) =>
	    javaInstance
	      .transaction('GET /api/product/list')
	      .duration(1000)
	      .success()
	      .timestamp(timestamp)
	      .serialize(),
	  );
const range = timerange(
  new Date('2021-01-01T00:00:00.000Z'),
  new Date('2021-01-01T00:15:00.000Z'),
);
const processor = new StreamProcessor({processors: [getSpanDestinationMetrics]});
processor.streamAsync(); //create async iterable including derived events
processor.streamToArray(); //helper to materialize to Array<> 
```

The biggest change to the `DSL` is that `rate()` no longer returns an array of timestamps on which we `flatMap` events. Instead it returns itself `Interval` which implements `Iterable<number>`. Using `spans()` we can describe an `SpanIterable` which implements `AsyncIterable<ApmFields[]>`. 


This PR also extends the CLI tool so that it works out of the box with `8.0.0` and cloud environments.


If you specify `--cloudId` the tool will now automatically migrate the cloud instance over to managed APM which in turn will setup all index/component templates needed for datastreams. 

You can now also explicitly specify the number of shards the backing indices for the datastreams should have. This will update the `*@custom` component templates for the apm datastreams. 

The `--clean` parameter when used in conjunction with `--cloudId` will now attempt to cleanly delete all apm datastreams. 

Lastly there is now a `--maxDocs` parameter that will break out of the event stream as soon as it processed the provided maximum of source events.

If provided this will force the interval generation to happen in a descending fashion from the `--to` parameter. This allows us to anchor any `recent N` queries in a sane fashion.

Mpdreamz committed:
Moves generator over to generators

* allow for lazy data generation
* feed lazy evaluation to a streamProcessor that does the aggregation.
* move entities over to use polymorphic this return types.
* introduces `SpanIterable` to aid with fluent construction

Mpdreamz committed:
Move to elasticsearch client bulk helper

and let it handle partitioning the data and concurrently index the data

Mpdreamz committed:
move new processors and hanlding into seperate files

Mpdreamz committed:
clean up StreamProcessor

Mpdreamz committed:
rename flatMap on SpanIterable to spans

Mpdreamz committed:
return datastreams always when calling getApmWriteTargets

Mpdreamz committed:
Add logger performance

Mpdreamz committed:
Update synthtrace ingestion to for 8.0.0

- Migrate cloud to managed out of the box
- Update to latest datastream mappings
- Move enrichment to StreamProcessor
- Instantiate services once, getCommonServices felt a bit service locator-ish

Mpdreamz committed:
Can now generate up to N spans

Mpdreamz committed:
Can now update the custom component templates

- For now we explicitly set number_of_shards to 1 (already the default)

Mpdreamz committed:
Remove check before swap, always swap

Mpdreamz committed:
update apm_api_integration usage of synthtrace lib

Mpdreamz committed:
fix compilation failures post-rebase

Mpdreamz committed:
make sure we generate transaction.root when calculating transaction metrics

Mpdreamz committed:
update timerange generation to generate bidirectionally properly again

elasticmachine commented:
Pinging @elastic/apm-ui (Team:apm)

Mpdreamz committed:
update EsLint errors

Commit:
commit using @elastic.co

Commit:
Address EsLint violations

Commit:
rename StackMonitoringFields back

Commit:
address TODO to move live indexation over to Syntthrace ES client

Mpdreamz commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into feature/synthtrace-stream-processing

Commit:
address post merge compilation error

dgieselaar commented:
@Mpdreamz did you happen to compare performance between the two different approaches?

Mpdreamz commented:
> @Mpdreamz did you happen to compare performance between the two different approaches?

Not in depth, this was mainly motivated to be more expressive not to fix an existing performance issue. Be good to know this does not introduce a regression though. 

I did check that memory consumption was stable when indexing a long time range.

dgieselaar commented:
@Mpdreamz Yeah, it was previously taking a few hours to fill a single 50gb shard. I want to make sure there is no significant performance regression (which may or may not be tied to architectural changes) before proceeding. Do you happen to already run it long enough to fill up a 50gb shard?

Mpdreamz committed:
Introduce --dryRun and include order on SpanIterable.

This now allows `StreamProcessor` to take into account streams that are
generated in both ascending and descending order e.g when `--maxDocs` is
specified.

Mpdreamz commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into feature/synthtrace-stream-processing

Mpdreamz committed:
StreamProcessor improvements

- add logger so it can report to the standard out.
- If multiple streams are passed it now interleaves the streams into a single stream.
  This ensures no single stream has preference

Mpdreamz committed:
Fixes incorrect categorization of transaction metrics as applicaton metrics.

dgieselaar commented on packages/elastic-apm-synthtrace/src/lib/apm/apm_fields.ts:
```diff
@@ -47,8 +52,7 @@ export type ApmFields = Fields &
     'metricset.name': string;
-    'observer.version': string;
-    'observer.version_major': number;
+    observer: Observer;
```
any reason why these are no longer flattened fields? (i'd prefer to keep it consistent)

Mpdreamz replied:
I was initially sending all data to Elasticsearch flattened but hit an in issue with the ingest pipelines we ship with: 

https://github.com/elastic/apm-server/issues/7088

Later I realize we `dedot` before sending data originally so included that back in. Will revert observer on `ApmFields` to be flat :+1:

dgieselaar commented on packages/elastic-apm-synthtrace/src/lib/apm/base_span.ts:
```diff
@@ -24,7 +24,7 @@ export class BaseSpan extends Serializable<ApmFields> {
   }
 
-  parent(span: BaseSpan) {
+  parent(span: BaseSpan): this {
```
is the explicit return type annotation necessary?

Mpdreamz replied:
Yes, this allows `parent()` to be called anywhere in the fluent chain and will always return the most concrete instance. Otherwise chaining to `parent()` will scope the DSL down to only those methods available on `BaseSpan`.

dgieselaar commented on x-pack/test/apm_api_integration/common/synthtrace_es_client_service.ts:
```diff
@@ -11,5 +11,6 @@ import { InheritedFtrProviderContext } from './ftr_provider_context';
 
-  return new apm.ApmSynthtraceEsClient(es, createLogger(LogLevel.info));
+  const forceDataStreams = false;
+  return new apm.ApmSynthtraceEsClient(es, createLogger(LogLevel.info), forceDataStreams);
```
At this point it might make sense to use a single argument with named keys rather than three separate ones.

Mpdreamz replied:
Cool happy to :smile:  Is there an official style guide recommendation for the under/over to move to `object` as args? I tend to consider `>3` args.

dgieselaar commented on packages/elastic-apm-synthtrace/src/lib/apm/processors/get_transaction_metrics.ts:
```diff
@@ -34,12 +34,13 @@ export function getTransactionMetrics(events: ApmFields[]) {
       return {
         ...transaction,
-        ['trace.root']: transaction['parent.id'] === undefined,
+        ['transaction.root']: transaction['parent.id'] === undefined,
```
thanks 😄

dgieselaar commented on packages/elastic-apm-synthtrace/src/lib/interval.ts:
```diff
@@ -6,27 +6,69 @@
-    private readonly to: number,
-    private readonly interval: string
-  ) {}
+    public readonly from: Date,
+    public readonly to: Date,
+    public readonly interval: string,
+    public readonly yieldRate: number = 1
```
this is probably better off at this point as a single argument as well

dgieselaar commented on packages/elastic-apm-synthtrace/src/lib/interval.ts:
```diff
@@ -6,27 +6,69 @@
+    public readonly interval: string,
+    public readonly yieldRate: number = 1
+  ) {
+    [this.intervalAmount, this.intervalUnit] = parseInterval(interval);
```
I think it's a little easier for the reader if we don't directly assign to class properties (e.g. by using an intermediate variable). I think most folks are not used to this approach.

Mpdreamz replied:
I wanted to localize the parsing of `amount` and `unit` to the constructor since its immutable state that otherwise `_generate()` needs to calculate every time a new enumeration is started.

dgieselaar commented on packages/elastic-apm-synthtrace/src/scripts/run.ts:
```diff
@@ -22,9 +24,23 @@ function options(y: Argv) {
+      describe: 'Elasticsearch target',
+      string: true,
+    })
+    .option('cloudId', {
+      describe:
+        'Provide connection information and will force APM on the cloud to migrate to run as a Fleet integration',
+      string: true,
+    })
+    .option('username', {
+      describe: 'Basic authentication username',
+      string: true,
       demandOption: true,
+    })
+    .option('password', {
+      describe: 'Basic authentication password',
       string: true,
+      demandOption: true,
     })
```
I think something like `--target=http://admin:changeme@localhost` is pretty easy to use, I'd rather not require separate authentication options. Can we make them optional and throw an error if cloudId is used instead of target? I think `yargs` also allows for either/or options w/ `argv.conflicts()` (for target/cloudId).

Mpdreamz replied:
Both should still work :)

dgieselaar commented on packages/elastic-apm-synthtrace/src/scripts/run.ts:
```diff
@@ -75,7 +105,9 @@ function options(y: Argv) {
-    .conflicts('to', 'live');
+    .conflicts('to', 'live')
+    .conflicts('maxDocs', 'live')
+    .conflicts('target', 'cloudId');
```
ah.. 😄

sorenlouv reviewed (approved):
Lots of good things happening in this PR! 

Without having a solid grasp on the implementation details I think this should be merged sooner rather than later, so we can be unblocked (existing Synthtrace doesn't work with data streams).

The removal of custom workers (and thus removal of parallelization) will impact perf but I don't think that's a dealbreaker. We should look into improving it in the future though.

sqren commented:
@Mpdreamz Can you please update the documentation for using Synthtrace in https://github.com/elastic/kibana/blob/main/x-pack/plugins/apm/dev_docs/local_setup.md#1-using-synthtrace ?

The suggestion is currently to use `es_archiver` to load the APM mappings. This is broken in 8.0 (see https://github.com/elastic/kibana/issues/112976#issuecomment-1047771035) and we want to move to the APM integration anyway. 

We can install the APM integration like:

```
curl $KIBANA_URL/api/fleet/epm/packages/apm/8.0.0 -X POST -H "Content-Type: application/json"  -H "kbn-xsrf: kibana" -d '{"force":true}'
```

I wonder if we can replace `8.0.0` with `latest` or something to like that, so we don't hardcode a specific version. @cauemarcondes / @ogupte do you know if this is possible?

ogupte commented:
> I wonder if we can replace 8.0.0 with "latest" or something to like that, so we don't hardcode a specific version. 

Not sure if the fleet kibana API currently supports a latest tag, but i do know they have a [fetchFindLatestPackage](https://github.com/elastic/kibana/blob/main/x-pack/plugins/apm/server/routes/fleet/get_apm_package_policy_definition.ts#L65) plugin API we use when we detect that kibana instance is a pre-release version. So it's possible.

Mpdreamz commented:
Thanks @ogupte I had registered a feature request for a `latest` moniker https://github.com/elastic/kibana/issues/121770. 


Currently the synth trace tool in this PR simply uses our own migrate API to automatically  switch over so there would not need to be a Step 2 anymore, will update this too @sqren :+1:

Mpdreamz committed:
Merge branch 'main' into feature/synthtrace-stream-processing

Mpdreamz committed:
fix new usages of synthtrace DSL

Mpdreamz committed:
fix eslint error

Mpdreamz committed:
move latency api integration over to new DSL

Mpdreamz committed:
address linter issues

Mpdreamz committed:
Merge branch 'main' into feature/synthtrace-stream-processing

Mpdreamz committed:
post merge update of DSL

Mpdreamz committed:
Merge branch 'main' into feature/synthtrace-stream-processing

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 123454` or prevent reminders by adding the `backport:skip` label.
